Switch to buildersyntax, create first composite

CollinValley opened the pull request:
Big PR,
Changes all of the links over to the new builder syntax. As well as does some renaming and housekeeping.

CollinValley committed:
AsyncLink: Move to builder pattern

Moves AsyncLink to builder pattern, and adds some additional required
trait bounds to element. This has to do with things requiring to be Send
when being put into a channel, and with the traits the compiler is no
longer able to infer that the E::Input and E::Output are Send. So we
state that as a fact.

CollinValley committed:
JoinLink: Move JoinLink to builder pattern

Only concern with this one is the static lifetime with packets, would
like to remove that somehow, but I am unsure what to anchor the lifetime
to.

CollinValley committed:
TeeLink: Switch to builder pattern

Commit switches TeeLink to builder pattern, no real concerns here
besides the 'static lifetime.

CollinValley committed:
Composite: Simple MtoN composite

CollinValley committed:
Rename TeeLink to CloneLink, remove builder from name

CollinValley committed:
Rename AsyncLink to QueueLink, remove builder from end of names

CollinValley committed:
BuilderConversion: Switch ClassifyLink to builder

This big commit mostly switches ClassifyLink to a builder syntax,
but it also handles some renaming and cleaning of the rest of the
runtime. Many of the files where named, formatted, or had their
type annotations set differently

CollinValley committed:
Remove uneccesary AsyncElement trait

CollinValley committed:
Run and cleanup fmt and clippy

ddowl commented on route-rs-runtime/src/element/identity_elements.rs:
```diff
@@ -17,7 +17,10 @@ impl<A> IdentityElement<A> {
 }
 
-impl<A> Element for IdentityElement<A> {
+impl<A> Element for IdentityElement<A>
```
Can this not have the type constraint `A: Send`?

CollinValley replied:
That seems reasonable, will change

CollinValley replied:
done

ddowl commented on route-rs-runtime/src/link/async_link.rs:
```diff
@@ -1,43 +1,98 @@
+        assert_eq!(
+            in_streams.len(),
+            1,
+            "AsyncLink may only take 1 input stream"
```
I may answer this later as I keep reading, but just confirming that we're keeping `Async` as 1 - 1, where we would use Link composition with `Join` and `Tee` in order to get the M - N effect?

CollinValley replied:
Exactly, which is why I changed it to queue link instead of async link

ddowl commented on route-rs-runtime/src/link/async_link.rs:
```diff
@@ -214,56 +269,105 @@ impl<Packet: Sized> Stream for AsyncEgressor<Packet> {
     use core::time;
     use futures::future::lazy;
 
+    fn run_tokio(runnables: Vec<TokioRunnable>) {
```
I'd like to consolidate these kinds of functions into test helpers, but I can do that after this PR is merged.

CollinValley replied:
I was thinking the same thing when I kept copying it lol

ddowl commented on route-rs-runtime/src/link/async_link.rs:
```diff
@@ -214,56 +269,105 @@ impl<Packet: Sized> Stream for AsyncEgressor<Packet> {
 
-        let link = AsyncLink::new(Box::new(packet_generator), elem, default_channel_size);
+        let (mut runnables, mut egressors) = AsyncLinkBuilder::new()
+            .ingressors(vec![Box::new(packet_generator)])
```
For each of these test examples, you can use the `ingressor` function you defined. `immediate_stream` is actually already inside a `Box`, so you can just use  `.ingressor(packet_generator)`

CollinValley replied:
I'm not sure that we should have an `ingressor` function at all. I think it kind of breaks our own trait concept? In a similar vein, for other functions, like braches, queue_length, etc, should we pull those into the trait and then make the links that don't need them implement them as no-ops?

ddowl replied:
I see it as a convenience method that isn't part of the `Link` trait. Something that users can specify when creating their own HoL (higher-order links) or composites, but obviously something we can't do in graphgen. As long as we're sure that

```builder.ingressor(in) == builder.ingressors(vec![Box::new(in)])```

then nothing is broken, and users get a slighly more ergonomic method to use.

CollinValley replied:
Hmm, put that way,  I do like having it there.  Might form a nice marker for the user to know that the link is 1 to stuff.

CollinValley replied:
Went ahead and implemented this for all links. On links that take multiple inputs, it works like a push() function. One links that only take one input, works as replace()

ddowl commented on route-rs-runtime/src/link/async_link.rs:
```diff
@@ -214,56 +269,105 @@ impl<Packet: Sized> Stream for AsyncEgressor<Packet> {
+    }
+
     #[test]
     fn async_link() {
```
nit: This test name should be more descriptive, or we can remove the example if it's not covering anything useful.

CollinValley replied:
doneish

ddowl commented on route-rs-runtime/src/link/join_link.rs:
```diff
@@ -1,51 +1,75 @@
-            task_parks.push(task_park);
+impl<Packet: Sized + Send + 'static> LinkBuilder<Packet, Packet> for JoinLinkBuilder<Packet> {
+    fn ingressors(self, in_streams: Vec<PacketStream<Packet>>) -> Self {
+        assert_ne!(in_streams.len(), 0, "Input stream vector can not be zero!");
```
Should we assert a rough upper bound on the number of ingress streams for sanity? I can't imagine a good design having more than 20 or so.

CollinValley replied:
I set this to 1000, decided to let people be dumb if they want. Left it at 1000 for ingress streams, egressors, and queue length. I doubt anybody wants numbers this high for anything else.

ddowl commented on route-rs-runtime/src/link/tee_link.rs:
```diff
@@ -1,50 +1,103 @@
-        assert!(
-            queue_capacity <= 1000,
-            format!("Tee Link queue_capacity: {} > 1000", queue_capacity)
+        assert_ne!(branches, 0, "branches must be non-zero");
```
Similar to `Join`, should we restrict the number of outgoing branches from `Tee`?

CollinValley replied:
Seems reasonable

CollinValley replied:
Fixed

ddowl commented on route-rs-runtime/src/link/tee_link.rs:
```diff
@@ -1,50 +1,103 @@
+    }
+}
+
+impl<Packet: Sized + Send + Clone + 'static> LinkBuilder<Packet, Packet>
```
Eventually the trait constraints on what can be passed through `Link`s might get a little out of hand, so maybe later we should alias those into another trait:

```
trait LinkPassable: Sized + Send + Clone + 'static
```

so we can write this definition as

```
impl<P: LinkPassable> LinkBuilder<P, P>
```

CollinValley replied:
Yeah, something like that.  I  want to talk about this in person. Since one we can't make packets 'static, and two we should reevaluate what our Trait requirements are in general.

CollinValley replied:
I did go through and remove the Sized requirement, since Clone requires Sized.

ddowl commented on route-rs-runtime/src/link/tee_link.rs:
```diff
@@ -1,50 +1,103 @@
 
-impl<P: Sized + Clone> TeeLink<P> {
-    pub fn new(input_stream: PacketStream<P>, queue_capacity: usize, branches: usize) -> Self {
+impl<Packet: Sized + Clone + Send> TeeLinkBuilder<Packet> {
```
We could also define an `ingressor` method for `TeeLink` that works similarly to other `Link`s that take just one ingressor.

CollinValley replied:
I feel like it is kludgy to have special methods in general for certain links,  would like to discuss with the group in person, see what everyone thinks.

CollinValley replied:
Done

ddowl commented on route-rs-runtime/src/composite/mton_composite.rs:
```diff
@@ -0,0 +1,157 @@
+    in_streams: Option<Vec<PacketStream<Packet>>>,
+    join_queue_capacity: usize,
+    tee_queue_capacity: usize,
+    branches: Option<usize>,
```
Is there a better name than `branches` to express the number of egressors the `Link` will produce? How about `num_egressors`?

CollinValley replied:
ack, good suggestion.

CollinValley replied:
Done

ddowl commented on route-rs-runtime/src/composite/mton_composite.rs:
```diff
@@ -0,0 +1,157 @@
+                .branches(self.branches.unwrap())
+                .build_link();
+            tee_runnables.extend(join_runnables);
+            (tee_runnables, tee_egresssors)
```
nit: Could we use `Vec::append` rather than `extend` to make it clear that we're returning all the runnables? I'd even prefer an extra name like:

```
let all_runnables = tee_runnables.append(join_runnables);
(all_runnables, tee_egressors)
```

CollinValley replied:
I suppose, I don't object to append.

CollinValley replied:
done

ddowl commented on route-rs-runtime/src/composite/mod.rs:
```diff
@@ -0,0 +1,4 @@
+/// Basic composite that take in M streams of a type, and outputs that same
+/// type to N packets.
+mod mton_composite;
```
nit: Since the job of this link isn't to really transform anything, just take M streams of a type and output copies to N streams, can we call it `MultiTee` or `MultiClone` or something? `composite` doesn't really mean anything on its own, and in fact we might want to hide the fact that any given Link is a composite since that's an implementation detail.

CollinValley replied:
Hmm I like MultiClone, or some other spiffy name, I agree the composite it kind of an impl detail.

ddowl commented on route-rs-runtime/src/lib.rs:
```diff
@@ -26,3 +26,7 @@ pub mod pipeline;
 /// Utility module
 mod utils;
+
+/// Composites are groups of links pre-assmebled to provide higher level functionality. They are highly customizable and users of the
+/// library are encourged to make their own to encourage code reuse.
```
:100:

ddowl commented on route-rs-runtime/src/link/mod.rs:
```diff
@@ -12,8 +12,8 @@ pub use self::sync_link::*;
 /// through the element defined process function, often performing some sort of transformation.
-mod async_link;
-pub use self::async_link::*;
+mod queue_link;
```
nit: I know that the queue structure is exposed via builder method, but if I'm a user I'm not quite sure what the "queue" gets me. This link can output to multiple egressors and wait for long-running processes to complete (which `SyncLink` can't),  but that's not obvious from the name `QueueLink`. I'm also not a big fan of `AsyncLink` either though. I'm cool with this for now, but we should definitely think hard about the names of our primitive links before any kind of release. We might notice some more API UX problems like, "if there's a SyncLink why is there no AsyncLink?", that might be more noticeable after finishing the NAT and having other people play around with the code. _user interviews??_

To put some more names into the ring: `BufferedLink` and `UnbufferedLink`. They're complements, and you can kindof intuit their properties based on the names?

CollinValley replied:
Hmm,  I do agree not having a Sync/Async complement is an issue. Not sure I like buffered link,   FYR, the queuelink is still 1-1.  If you want multi-output, then you need a CloneLink.   I thought about supporting the queuelink as multi-output, but figured it would be better to just put a SyncLink->CloneLink to get the desired effect at minimal cost.  

Crazytown thought. What if the only Link that can take elements is the SyncLink?  QueueLink or BufferLink is for...buffers?  We could rename Clone to Fork, so there are obvious Fork/Join complements.  Classify would remain its own special link

ddowl replied:
https://gph.is/g/E3geJNn

CollinValley replied:
Going to leave this thought for meetup. We should make the choice as a group, since it basically will cement the design philosophy for our links overall.

ddowl commented on route-rs-runtime/src/link/classify_link.rs:
```diff
@@ -1,64 +1,140 @@
-            queue_capacity <= 1000,
-            format!("ClassifyLink queue_capacity: {} > 1000", queue_capacity)
+            (2..=1000).contains(&branches),
+            format!("Branches: {}, must be in range 2..=1000", branches)
```
"must be between 2 and 1000 branches"?

CollinValley replied:
At first I thought I would protect the user from making a 1-1 classify cause what is the point? Same for 1-1 CloneLink. Decided to remove that cause it isn't completely the libraries job to protect the user from stupid inefficiency. There may be a use we aren't considering.

ddowl commented on route-rs-runtime/src/link/classify_link.rs:
```diff
@@ -1,64 +1,140 @@
-        dispatcher: Box<dyn Fn(C::Class) -> usize + Send + Sync + 'a>,
-        queue_capacity: usize,
-        branches: usize,
+impl<C: Classifier> ClassifyLink<C> {
```
We can also define an `ingressor` function on `ClassifyLink` that accepts only 1 `PacketStream`.

CollinValley replied:
done

ddowl commented on route-rs-runtime/src/link/classify_link.rs:
```diff
@@ -167,37 +244,97 @@ mod tests {
-            number_branches,
-        );
+        let (mut runnables, mut egressors) = ClassifyLink::new()
+            .ingressors(vec![Box::new(packet_generator)])
```
`ClassifyLink`'s test examples can also use the simpler `ingressor` builder function as well.

CollinValley replied:
done

ddowl commented on route-rs-runtime/src/link/mod.rs:
```diff
@@ -34,9 +34,9 @@ pub use self::blackhole_link::*;
+pub type PacketStream<Input> = Box<dyn Stream<Item = Input, Error = ()> + Send>;
 /// Some Links may need to be driven by Tokio. This represents a handle to something Tokio can run.
-pub type TokioRunnable = Box<dyn Future<Item = (), Error = ()> + 'static + Send>;
+pub type TokioRunnable = Box<dyn Future<Item = (), Error = ()> + Send>;
```
Is removing the static lifetime on these types intentional?

CollinValley replied:
Sorta, patched this.

CollinValley committed:
Runtime: Add 'static lifetime to TokioRunnable

Tokio requires that its runnables be static, so we require that the
things we make TokioRunnables out of also be static.

CollinValley committed:
BlackholeLink: Switch to Element from Packet

Previously, we had an issue that the we were making Blackhole link generic
off the Packet type that is flowing into it as input. This instictually makes
sense, but due to our need to create TokioRunnables with 'static lifetimes, it
means that the Packet would have to be 'static. A way to get around this is to
use an IdentityElement which is 'static

CollinValley committed:
Links: Rework asserts to be more clear

Rework asserts to be more consise and clear, and remove a few that
were not needed. Also allow user to create CloneLinks of branch amount 1,
this is silly for them to do but there is no reason to stop them.  Also
use the variable num_egressors instead of branches to be more clear for the
reader.

CollinValley commented:
Recent changes.

- Resolved weirdness with Asserts, they are all much more consistent
- BlackholeLink: Tried to eliminate the 'static packet,  by requiring a phantom Element.  This is less than ideal, but I think it works to fix our issue.  If we go down this route, we would likely make the element real rather than a dummy.
- 'static lifetime: Fixed the error where I removed the 'static requirement from TokioRunnable, left it off on PacketStream?  It doesn't seem to cause an issue.

More work still to do, but this is what I could do on the plane with no internet.

CollinValley committed:
Composite: Create m_transform_n_composite

Create an example composite. This one takes in M ingress streams
of a certain type, Input, an Element loaded into a syncLink that
transforms Input->Output. Then user can also specify how many egressors
get copies of the Output streams.

CollinValley committed:
CloneLink: Require that packets are Clone

We need to be able to Clone packets for CloneLink and Classify, so require
that the input and output types of Element implement Clone.

CollinValley committed:
Composite: M transform N

This composite takes M streams of type Input, uses a provided element
that maps Input->Ouput, and then provides N egressors that implement
Stream of type Output. Pretty good example of the power of
composite links. Not completely sold on the name

CollinValley committed:
queue_link: Make basic test more descriptive

CollinValley committed:
Links: Add ingressor function

Add ingressor function to all links. On links that only take one
input stream max, replace the current input_stream. On links that
take more than one, add the provided input_stream to the list of
input streams.

CollinValley committed:
Link: Use Vec::append rather than extend

Drew requested this, I don't really have an issue with doing it this
way, it does seem a bit more clear that we are returning all the
runnables.

CollinValley committed:
Clippy fixup

ddowl commented on route-rs-runtime/src/link/queue_link.rs:
```diff
@@ -327,7 +327,7 @@ mod tests {
 
     #[test]
-    fn queue_link() {
+    fn queue_link_works() {
```
HAHAHAHA

ddowl commented:
LGTM. Let's try and make PRs a little smaller in the future.

ddowl reviewed (approved):
Whoops, meant to approve.
